.github/workflows/test_on_push.yml

```diff
@@ -114,18 +142,46 @@ jobs:
         if: matrix.os == 'ubuntu-latest' && matrix.python-version == 3.11
         uses: codecov/codecov-action@v2.1.0
 
+      - name: Cache integration tests nox environment for GNU/Linux with Python 3.11
+        uses: actions/cache@v3
+        if: matrix.os == 'ubuntu-latest' && matrix.python-version == 3.11
+        with:
+          path: ${{ github.workspace }}/.nox/integration/
+          key: ${{ runner.os }}-nox-integration-${{ matrix.python-version }}-${{ hashFiles('**/noxfile.py', '**/setup.py') }}
+
       - name: Run integration tests for GNU/Linux with Python 3.11
         if: matrix.os == 'ubuntu-latest' && matrix.python-version == 3.11
         run: nox -s integration
 
-      - name: Run unit tests for Windows and MacOS with all Python versions
+      - name: Cache unit tests nox environment for Windows and macOS with all Python versions
+        uses: actions/cache@v3
+        if: matrix.os != 'ubuntu-latest'
+        with:
+          path: ${{ github.workspace }}/.nox/unit/
+          key: ${{ runner.os }}-nox-unit-${{ matrix.python-version }}-${{ hashFiles('**/noxfile.py', '**/setup.py') }}
+
+      - name: Run unit tests for Windows and macOS with all Python versions
```

Comment:
Is it required to cache ubuntu and (mac + windows) dependencies separately?

Unrelated to this PR, but we thanks to sing nox sessions for multiple OS, we might be able to refactor this workflow file a bit. Could we then cache everything in a single step?

Reply:
Yes, we can definitely do it in a single step—it might even be faster because it would then perform the copying operation for the `nox` environments just once.  Should we cache the `pybamm-requires` + related files in one step, and the `unit`, `integration`, `examples`, and `doctests` environments in another step? Or should we group caching by OS?

Reply:
> Should we cache the pybamm-requires environment + related files in one step, and the unit, integration, examples, and doctests environments in another step.

I think this sounds good, given that I am trying to get rid of OS separation in the workflows - #3095

Reply:
I was going to try this but I realised that we have different files to hash for different environments, so we will have to keep three separate cache steps: one that caches `unit` + `integration` + `examples`, one for `pybamm-requires`, and one for `doctests` (with different sets of `extras` installations)

I think it would be better to keep all environments' caches separate (i.e., as they currently are) since they contain the name of the environment too, which will make debugging easy—unless you have a better idea